experiments/mountain_car_plots.py

```diff
@@ -0,0 +1,269 @@
+
+import gym
+from matplotlib import pyplot as plt
+import numpy as np
+import pandas as pd
+import seaborn as sns
+from stable_baselines.common.vec_env import DummyVecEnv, VecNormalize
+
+from imitation import util
+from imitation.policies.serialize import load_policy
+from imitation.rewards.serialize import load_reward
+from imitation.util.reward_wrapper import RewardFn
+
+MC_POS_MIN, MC_POS_MAX = -1.2, 0.6
+MC_VEL_MIN, MC_VEL_MAX = -0.07, 0.07
+MC_GOAL_POS = 0.5
+
+
+# Utility for calculating the next observation s'.
+# Required for evaluating AIRL reward at arbitrary (s, a) points.
+def make_next_obs(obs, acts) -> np.ndarray:
+  """Utility for calculating the next observation s'.
+
+  Required for evaluating AIRL reward at arbitrary (s, a) points.
+  """
+  env = gym.make("MountainCar-v0")
+  obs = np.array(obs)
+  acts = np.array(acts)
+  next_obs = []
+  for ob, act in zip(obs, acts):
+    assert ob.shape == (2,)
+    env.reset()
+    env.env.state = ob
+    next_ob, *_ = env.step(act)
+    next_obs.append(next_ob)
+  return np.array(next_obs)
+
+
+def make_policy_rollout(policy_path, env_name="MountainCar-v0",
+                        ) -> List[util.rollout.Trajectory]:
+  """Load policy from path and return a list of Trajectories."""
+  venv = DummyVecEnv([lambda: gym.make(env_name)])
+  with load_policy("ppo2", str(policy_path), venv) as gen_policy:
+    trajs = util.rollout.generate_trajectories(
+      gen_policy, venv, sample_until=util.rollout.min_episodes(50))
+  return list(trajs)
+
+
+def make_heatmap(
+    act: int,
+    reward_fn: RewardFn,
+    pos_step_size: float = 0.1,
+    velocity_step_size: float = 1e-2,
+    mark_goal: bool = True,
+    gen_trajs: Optional[List[util.rollout.Trajectory]] = None,
+    exp_trajs: Optional[List[util.rollout.Trajectory]] = None,
+    legend_on: bool = True,
+    title: bool = None,
+    heatmap: bool = True,
+    filter_trans_by_act: bool = True,
+) -> None:
+  """
+  Make a MountainCar heatmap of rewards for a particular action. X axis
+  is position. Y axis is velocity.
+
+  Args:
+    act: The MountainCar action number whose reward we are evaluating.
+      Should be 0, 1, or 2.
+    reward_fn: Reward function. Should accept unnormalized inputs.
+    pos_step_size: The x axis granularity.
+    velocity_step_size: The y axis granularity.
+    mark_goal: If True then plot a line for goal position.
+    gen_trajs: A list of generator trajectories to
+      scatterplot on top of the heatmap.
+    exp_trajs: A list of exp trajectories to scatterplot on
+      top of the heatmap.
+    legend_on: Whether to plot the legend.
+    title: Custom title.
+    heatmap: Whether to plot the heatmap.
+    filter_trans_by_act: If True, then filter out transitions from
+      `gen_trajs` and `exp_trajs` that don't use action `act` before
+      scatterplotting.
+  """
+  assert 0 <= act <= 2
+
+  obs_vec, acts_vec = [], []
+  for p in np.arange(MC_POS_MIN, MC_POS_MAX + 1e-5, step=pos_step_size):
```

Comment:
How about specifying the number of data points for position and velocity, and then using `np.linspace`? This will guarantee the endpoints get included, and avoid needing the `+ 1e-5` trick.

Reply:
This suggestion becomes necessary now that I've moved from `seaborn.heatmap` to `matplotlib.pcolor(X, Y, Z)`.